tools/mason/MasonPublish.chpl

```diff
@@ -0,0 +1,248 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use MasonUtils;
+use Spawn;
+use FileSystem;
+use TOML;
+use MasonEnv;
+use MasonNew;
+use MasonModify;
+
+/* Top Level procedure that gets called from mason.chpl that takes in arguments from command line
+Returns the help output in '-h' or '--help' exits in the arguments.
+If --dry-run is passed then it checks to see if the package is able to be published.
+Takes in the username of the package owner as an argument*/
+proc masonPublish(args: [] string) throws {
+  try! {
+    if args.size < 3 {
+      masonPublishHelp();
+      exit(0);
+    }
+    var dry = false;
+    var username = "";
+    if args.size == 3 || args.size == 4 {
+      for arg in args[2..] {
+        if arg == '-h' || arg == '--help'{
+          masonPublishHelp();
+          exit(0);
+        }
+        else if arg == "--dry-run" {
+          dry = true;
+        }
+        else {
+          username = arg;
+        }
+      }
+      if doesGitOriginExist() {
+        if dry && args.size == 4 {
+          dryRun(username);
+        }
+        else if !dry && args.size == 4 {
+          writeln('"' + args[0]+ ' '  + args[1] + ' ' + args[2] + ' ' + args [3] +'" does not meet the "mason publish [options] <username>" syntax');
+          writeln('See "mason publish -h" for more details');
+          exit(0);
+        }
+        else if args.size == 3 && !dry {
+          var usernameCheck = ('git ls-remote https://github.com/' + username + '/mason-registry').split();
+          var p = spawn(usernameCheck);
+          p.wait();
+          if p.exit_status == 0 {
+            publishPackage(username);
+          }
+          else {
+                throw new owned MasonError(username + ' is not a valid GitHub username or mason-registry is not forked');
+          }
+        }
+        else if args.size == 3 && dry {
+          throw new owned MasonError('Must pass your username with --dry-run');
+        }
+      }
+      else {
+        throw new owned MasonError('Must have remote origin set up in repository in order to publish.');
+      }
+    }
+    else throw new owned MasonError('Usage must meet the form "mason publish [options] <username>"');
+  }
+  catch e: MasonError {
+    writeln(e.message());
+    exit(1);
+  }
+}
+
+/* Main Script that goes through the act of publishing the package to the mason registry.
+ Takes the package owners GitHub usrename as input will throw errors through command
+line git commands if any of the git calls fails.*/
+proc publishPackage(username: string) throws {
+  try! {
+    const packageLocation = getEnv('PWD');
+    here.chdir(MASON_HOME);
+    const dir =  getEnv('PWD');
+    mkdir('tmp');
+    here.chdir(MASON_HOME + '/tmp');
```

Comment:
What if multiple mason clients are being run at the same time? 

Mason is not parallel safe today, but we might as well make this operation parallel-safe to reduce the amount of work needed in the future. I'd suggest attaching a unique ID to this tmp directory name. A random integer might be safe enough for now, e.g. `MASON_HOME/tmp/uniqueID`.